### A Layer of Neurons

 Layer of Neurons Neural networks typically have layers that consist of more than one neuron. Layers are nothing more than groups of neurons. Each neuron in a layer takes exactly in the input given to the layer before it (which can be either the training data or the output from the previous layer). 

 Let’s say we have a scenario with 3 neurons in a layer and 4 inputs:

 ![A Layer of Neurons](./img/diagram%202.png)

In [3]:
inputs = [1, 2, 3, 2.5] 

weights1 = [0.2, 0.8, -0.5, 1] 
weights2 = [0.5, -0.91, 0.26, -0.5] 
weights3 = [-0.26, -0.27, 0.17, 0.87]

bias1 = 2 
bias2 = 3 
bias3 = 0.5

outputs = [ 
    # Neuron 1: 
    inputs[0]*weights1[0] + 
    inputs[1]*weights1[1] + 
    inputs[2]*weights1[2] + 
    inputs[3]*weights1[3] + 
    bias1,
    
    # Neuron 2: 
    inputs[0]*weights2[0] + 
    inputs[1]*weights2[1] + 
    inputs[2]*weights2[2] + 
    inputs[3]*weights2[3] + 
    bias2,
    
    # Neuron 3: 
    inputs[0]*weights3[0] + 
    inputs[1]*weights3[1] + 
    inputs[2]*weights3[2] + 
    inputs[3]*weights3[3] + bias3
]

print(outputs)

[4.8, 1.21, 2.385]


Outputs should be **4.8, 1.21, 2.385**

![Output of Layer](./img/diagram%203.png)

Changing the code to use loops instead of the hardcoded operations, and turning the separate weight variables into a list of weights so we can iterate over them:

In [4]:
inputs = [1, 2, 3, 2.5] 
weights = [
    [0.2, 0.8, -0.5, 1], 
    [0.5, -0.91, 0.26, -0.5], 
    [-0.26, -0.27, 0.17, 0.87]
] 
biases = [2, 3, 0.5]

# Output of current layer 
layer_outputs = []
# For each neuron
for neuron_weights, neuron_bias in zip(weights, biases):
    # Zeroed output of given neuron
    neuron_output = 0
    # For each input and weight to the neuron 
    for n_input, weight in zip(inputs, neuron_weights):
        # Multiply this input by associated weight and add to the neuron’s output
        neuron_output += n_input*weight
    # Add bias 
    neuron_output += neuron_bias 
    # Put neuron’s result to the layer’s output list 
    layer_outputs.append(neuron_output)
    
print(layer_outputs)

[4.8, 1.21, 2.385]
